Implemented per-collection thread runtime metrics

jbaramidze opened the pull request:
At the moment, if you want to check the CPU usage of various threads by thread analyzer subsystem, you can only get average CPU usage since boot. This PR implements mechanism to maintain and fetch CPU usage between metric collection events. This way you can have much better insights at how CPU usage changes over time.

peter-mitsis commented on kernel/Kconfig:
```diff
@@ -395,6 +395,14 @@ config SCHED_THREAD_USAGE
+	default n
+	depends on SCHED_THREAD_USAGE
+	help
+ 	  Enable collection of runtime statistics between metri collections.
```
Typo: metri -> metrics

peter-mitsis reviewed (commented):
Aside from the coding style items that have been auto-flagged (which need to be fixed), this looks on first pass to be decent.

Is it possible to shorten the field name "utilization_since_last_collection"? Perhaps "utilization_delta"? (This is not a make or break suggestion.)

The only other suggestion at this point I have is centres around organising the commits. When possible, I find it helpful to see new functionality split into different categories: implementation, usage and test(s). Again, this is not a make or break--it just helps my poor brain keep track of the various details.

I am looking forward to re-reviewing this.

jbaramidze committed:
kernel: thread: Implemented per-collection runtime metrics gathering

Added new fields to thread statistics that can be reset after
reading. This allows us to calculate things like CPU usage during
last X seconds by resetting the new fields and then checking them
after X seconds.

Signed-off-by: Zhani Baramidze <jbaramidze@gmail.com>

jbaramidze committed:
debug: thread_analyzer: Use per-collection metrics

Extend the thread_analyzer subsystem to use newly
added per-collection metrics.

Signed-off-by: Zhani Baramidze <jbaramidze@gmail.com>

jbaramidze committed:
tests: Test per-collection metrics

Added bunch of test to test the per-collection
metrics functionality.

Signed-off-by: Zhani Baramidze <jbaramidze@gmail.com>

andyross reviewed (dismissed):
Looks correct, though I'll fully admit that thread_analyzer is one of the corners of the kernel API about which I know the least.  Treat @peter-mitsis's review here as more authoritative.

kartben commented:
@jbaramidze Will you have time to look at fixing the coding style issues? Also, looks like you will also need to rebase on top of main and resolve conflict in `include/zephyr/kernel/stats.h`. Thanks again for the PR!

jbaramidze commented:
@kartben @andyross  @peter-mitsis 
Sorry for the delay guys. Fixed all the mentioned issues and pushed. Looking forward to your re-review!

peter-mitsis commented on kernel/Kconfig:
```diff
@@ -464,6 +464,14 @@ config SCHED_THREAD_USAGE
+	depends on SCHED_THREAD_USAGE
+	help
+	  Enable collection of runtime statistics between metrics collections.
+	  For example, Thread exection cycles since last time it got collected.
```
"Thread exection" -> "thread execution"

jbaramidze replied:
Fixed!

peter-mitsis reviewed (commented):
Aside from a minor typo in the Kconfig help for THREAD_RUNTIME_STATS_BETWEEN_COLLECTIONS, this looks good.

jbaramidze commented:
A couple of tests failed, but the failures seem unrelated. They didn't fail before my last comment fix.

cfriedt reviewed (approved):
LGTM - in terms of SMP, is there anything special that will be required? How would the test suite differ?

andyross commented on kernel/thread.c:
```diff
@@ -1167,14 +1167,36 @@ int k_thread_runtime_stats_all_get(k_thread_runtime_stats_t *stats)
 	return 0;
 }
+
+void k_thread_runtime_stats_all_cleanup(void)
```
New API functions need docs.  In particular ones introducing new features.

Also not really loving the name.  "Clean up" implies that there is some garbage or something that needs to be tidied, and generally is used for steps that have to be done at some point.  This is closer to, maybe, "checkpoint": you're tagging a specific moment in the statistics data you want to use later.

andyross reviewed (commented):
Definitely needs docs for the new API (doesn't need to be complicated, but the doxygen output should show at least something) ~~and a test for coverage, and maybe a quick example in a sample somewhere (the example and test could be the same code, this is really simple).~~  (Sorry, had clicked out only the first patch, there's a test there, never mind)

carlescufi commented:
@jbaramidze can you follow-up on @andyross's review comments please?

nashif commented on kernel/Kconfig:
```diff
@@ -464,6 +464,14 @@ config SCHED_THREAD_USAGE
+	default n
+	depends on SCHED_THREAD_USAGE
+	help
+	  Enable collection of runtime statistics between metrics collections.
```
Wording here is awkward, enable collection between collections?
Same for the Kconfig itself.

cfriedt commented:
From a practical perspective, we should probably implement per thread cpu usage as an up-counter of some kind.

Callers should perform any averaging / windowing / instantaneous calculations since that will always be relative to the observer and therefore not really possible to manage coherently for multiple callers.
